dask/array/slicing.py

```diff
@@ -559,9 +559,14 @@ def slicing_plan(chunks, index):
 
     if not is_arraylike(index):
         index = np.asanyarray(index)
-    cum_chunks = cached_cumsum(chunks)
 
-    cum_chunks = asarray_safe(cum_chunks, like=index)
+    cum_chunks_tup = cached_cumsum(chunks)
+    cum_chunks = asarray_safe(cum_chunks_tup, like=index)
+    if cum_chunks.dtype.kind != "f":  # Don't cast NaN chunks to int
+        # This is important when index.dtype=uint64 (or uint32 on 32-bit hosts) to
+        # prevent accidental automatic casting during `index - cum_chunks` below
+        cum_chunks = cum_chunks.astype(index.dtype)
```

Comment:
cum_chunks was int64 before (in numpy 1 too); this PR changes it to uint64 (or whatever the end user passed as a slicer).
This fixes the dtype arithmetics on line 586 below, which used to subtract uint64 - int64, which in numpy 1 returns uint64 but in numpy 2 returns float.